app/src/main/kotlin/com/wire/android/feature/auth/login/LoginUseCase.kt

```diff
@@ -0,0 +1,14 @@
+package com.wire.android.feature.auth.login
+
+import com.wire.kalium.logic.CoreLogic
+import javax.inject.Inject
+import com.wire.kalium.logic.feature.auth.AuthenticationResult
+import dagger.hilt.android.scopes.ViewModelScoped
+
+@ViewModelScoped
+class LoginUseCase @Inject constructor(private val coreLogic: CoreLogic) {
+    suspend operator fun invoke(userIdentifier: String, password: String): AuthenticationResult =
+        coreLogic.authenticationScope {
+            login(userIdentifier, password, true)
```

Comment:
In this case we will have usecases in both android reloaded and kalium - in the documentation on the graph it looks like they are only supposed to be in kalium - are we changing this concept?

Would it be worse to inject and use `corelogic` directly in `LoginViewModel`?

Reply:
the issue with using `CoreLogic` is `ViewModel` will become harder to understand since almost all of the view models will look like this:
```kotlin
@HiltViewModel
class MyViewModel @Inject constructor(
    private val coreLogic: CoreLogic
) : ViewModel() {
```
instead of a more descriptive:
```kotlin
@HiltViewModel
class MyViewModel @Inject constructor(
    private val getMessagesUseCase: GetMessagesUseCase,
    private val updateMessageUsecase: UpdateMessageUsecase
) : ViewModel() {
```
and it makes it a bit harder to understand what a viewModel dose from the first look

Reply:
Can't we inject Kalium's usecases instead of having this tiny eggshell around it?

Reply:
@vitorhugods @saleniuk i did some changes to the hilt `Module`
i was able to inject the `Logic` use cases 

```kotlin
@Module
@InstallIn(SingletonComponent::class)
class CoreLogicModule {

    @Singleton
    @Provides
    fun coreLogicProvider(@ApplicationContext context: Context): CoreLogic {
        ...
        return CoreLogic(.....)
    }

    @Singleton
    @Provides
    fun loginUserCaseProvider(coreLogic: CoreLogic): LoginUseCase = coreLogic.getAuthenticationScope().loginUsingEmail
}

```

Reply:
this approach will not work with other use cases since `getSessionScope(session: AuthSession)` needs session.

Reply:
Yeah.

The ones that require a session are troublesome. And can be split based on their nature I guess. Things like `GetSelfUserProfileUseCase` or similar are only used for the `current` session, _i.e._ the session that is currently selected and that the UI is displaying. So we could have something like:
```kotlin
@Factory
@Provides
fun getSelfUserProfileUseCase(coreLogic: CoreLogic, sessionManager: SessionManager): GetSelfUserProfileUseCase 
    = coreLogic.getSessionScope(sessionManager.currentSession).user.selfProfileData
```

Things like `SendMessageUseCase` are a bit harder. Because technically we could send messages when the user replies in a system notification of a new message from another session, different from the one being displayed in the UI.

Well, I don't know. Should we support this initially? Maybe it's something we need to discuss with the designers.
In any case, if we don't support initially and use only the `currentSession` for everything, it will suck big time if we need to refactor later.

Reply:
In such case, the message that is shown as a notification should contain id of the user this message is sent to, so maybe we could have access to other sessions, like `sessionManager.getSessionForUserId(userId)`?

Reply:
@saleniuk exactly. But how to tie that with injection I have no idea yet :P

Reply:
`SessionManager` can have two methods and be available from `CoreLogic` or even add these two methods directly to the `CoreLogic`
```
    fun getCurrentAuthSession(): AuthSession?
    fun getAuthSessionForUserId(userId: String): AuthSession?
```
but I don't know what you think about sharing `AuthSession` outside kalium, maybe `coreLogic.getSessionScope` should take `userId` as a parameter and by default it should be `currentAuthSession.userId`

Reply:
We are adding persistence to Kalium, and I guess we can manage the storage of multiple-sessions there, yes. 

But still, the app at some point will have to choose which session is doing the operation.
Some (probably most) will be for the current session. For that reason injecting the `UseCase` works perfectly by defining a `UseCase` as `CoreLogic` + `currentSession`.
But for some... It just can't be done and must be resolved depending on which session is receiving a call, or replying to a message through notifications, etc.

So, maybe we can start by adding `@ViewModelScoped` to all things that rely on `currentUser`. We scope down this definition to only be used with ViewModels. This means that the view must be visible, and we're dealing with the current user. Kinda makes sense.

When we need to handle system notifications or some kind of event that starts in the background, we can figure this out later. Worst case scenario we use `CoreLogic` directly in these elements.

Reply:
If I understand it correctly, now we have to make `coreLogic. getSessionScope` without parameter, let it return the `SessionScope` for the `currentUser` and keep the `currentUser` only inside the kalium?